sound/soc/sof/intel/hda-codec.c

```diff
@@ -163,6 +163,8 @@ int hda_codec_i915_exit(struct snd_sof_dev *sdev)
 	struct hdac_bus *bus = sof_to_bus(sdev);
 	int ret;
 
+	/* balance of snd_hdac_ext_bus_device_init */
+	snd_hdac_ext_bus_device_remove(bus);
```

Comment:
Yes, this is one of the issues, but I don't think it belongs here, this is more of a generic code that isn't directly tied to i915. Should this go in the .remove callback instead?

Reply:
@plbossart Thanks for pointing it out. I am also thinking about it. I wrote it here is because I think it is better if the both functions are in the same file. snd_hdac_ext_bus_device_init is called in hda-codec.c and hda_codec_i915_exit() is only called by hda_dsp_remove().

Reply:
yes, this should go to hda.c:hda_dsp_remove().

Reply:
> @plbossart Thanks for pointing it out. I am also thinking about it. I wrote it here is because I think it is better if the both functions are in the same file. snd_hdac_ext_bus_device_init is called in hda-codec.c and hda_codec_i915_exit() is only called by hda_dsp_remove().

Hi Bard, does the sequence below help to undertand?

hda_dsp_probe()
    hda_init()
    hda_init_caps()
        hda_codec_i915_init()
        hda_codec_probe_bus()
            hda_codec_probe()
                snd_hdac_ext_bus_device_init()

hda_dsp_remove()
    hda_deinit_caps() //virtual, not exist
        hda_codec_remove_bus() //virtual, not exist
            hda_codec_remove() //virtual, not exist
                snd_hdac_ext_bus_device_remove()
        hda_codec_i915_exit()
    hda_deinit() //virtual, not exist

Reply:
@keyonjie Thanks. it's clear.  But should it be inverse sequence? ie. hda_codec_i915_exit() first and snd_hdac_ext_bus_device_remove() latter.

Reply:
@bardliao if we look at the SKL driver, it's snd_hdac_ext_bus_device_remove and i915_exit later. It makes sense, the i915 part is to enable the graphics link, the hda functions are to interact with the codec exposed as a result of enabling the i915 connection.

I am still unclear if we also need to call snd_hdac_ext_stop_streams() snd_hdac_ext_bus_exit() as well (and all the snd_hdac_bus_free_stream_pages snd_hdac_stream_free_all snd_hdac_link_free_all). I'd really like to have a call tree where we can show how the hdac_ and hdac_ext_ functions are used.

Reply:
@plbossart I moved it to the front of platform_device_unregister() by referring to skl_remove(). Also, I added snd_hdac_link_free_all() and snd_hdac_ext_bus_exit().  snd_hdac_link_free_all() will free hlink_list which is added by snd_hdac_ext_bus_get_ml_capabilities.
```
hda_dsp_probe()
         hda_init_caps()
                 snd_hdac_ext_bus_get_ml_capabilities()
                         list_add_tail(&hlink->list, &bus->hlink_list)
```
snd_hdac_ext_bus_exit() is balanced with snd_hdac_ext_bus_init() and will cancel unsol_work.
We don't use snd_hdac_ext_bus_init() but we do similar thing in sof_hda_bus_init().
```
snd_hdac_ext_bus_exit()
        snd_hdac_bus_exit()
                cancel_work_sync(&bus->unsol_work)
```
```
hda_dsp_probe()
        hda_init()
                sof_hda_bus_init()
                        INIT_WORK(&bus->unsol_work, snd_hdac_bus_process_unsol_events);
```

After adding these, running sof_bootone.sh is more stable than before. But sometimes I will get 
```
RIP: 0010:kmem_cache_alloc_trace+0x93/0x1e0
```
I think it is from memory allocation. So, I think there are still something we didn't free.
I will move the discussion to https://github.com/thesofproject/linux/issues/299